trie/stacktrie.go

```diff
@@ -0,0 +1,384 @@
+
+		// Finally, cut off the key part that has been passed
+		// over to the children.
+		st.key = st.key[:diffidx]
+		st.val = nil
+	case emptyNode: /* Empty */
+		st.nodeType = leafNode
+		st.key = key[st.keyOffset:]
+		st.val = value
+	case hashedNode:
+		panic("trying to insert into hash")
+	default:
+		panic("invalid type")
+	}
+}
+
+func (st *StackTrie) hash() []byte {
+	/* Shortcut if node is already hashed */
+	if st.nodeType == hashedNode {
+		return st.val
+	}
+
+	var preimage bytes.Buffer
+	var n node
+	d := sha3.NewLegacyKeccak256()
+	switch st.nodeType {
+	case branchNode:
+		fn := &fullNode{}
+		payload := [544]byte{}
+		pos := 3 // maximum header length given what we know
+		for i, v := range st.children {
+			if v != nil {
+				// Write a 32 byte list to the sponge
+				childhash := v.hash()
+				if len(childhash) == 32 {
+					payload[pos] = 128 + byte(len(childhash))
+					pos++
+				}
+				copy(payload[pos:pos+len(childhash)], childhash)
+				pos += len(childhash)
+				if len(childhash) < 32 {
+					fn.Children[i] = rawNode(childhash)
+
+				} else {
+					fn.Children[i] = hashNode(childhash)
+				}
+				st.children[i] = nil // Reclaim mem from subtree
+			} else {
+				// Write an empty list to the sponge
+				payload[pos] = 0x80
+				pos++
+			}
+		}
+		// Add an empty 17th value
+		payload[pos] = 0x80
+		pos++
+
+		// Compute the header, length size is either 0, 1 or 2 bytes since
+		// there are at least 17 empty list headers, and at most 16 hashes
+		// plus an empty header for the value.
+		var start int
+		if pos-3 < 56 {
+			payload[2] = 0xc0 + byte(pos-3)
+			start = 2
+		} else if pos-3 < 256 {
+			payload[2] = byte(pos - 3)
+			payload[1] = 0xf8
+			start = 1
+		} else {
+			payload[2] = byte(pos - 3)
+			payload[1] = byte((pos - 3) >> 8)
+			payload[0] = 0xf9
+			start = 0
+		}
+
+		// Do not hash if the payload length is less than 32 bytes
+		if pos-start < 32 {
+			// rlp len < 32, will be embedded
+			// into its parent.
+			return payload[start:pos]
+		}
+		n = fn
+		preimage.Write(payload[start:pos])
+	case extNode:
+		ch := st.children[0].hash()
+		n = &shortNode{
+			Key: hexToCompact(st.key),
+			Val: hashNode(ch),
+		}
+		err := rlp.Encode(&preimage, n)
+		st.children[0] = nil // Reclaim mem from subtree
+		if err != nil {
+			panic(err)
+		}
+		if preimage.Len() < 32 {
+			return preimage.Bytes()
+		}
+	case leafNode:
+		n = &shortNode{
+			Key: hexToCompact(append(st.key, byte(16))),
+			Val: valueNode(st.val),
+		}
+		err := rlp.Encode(&preimage, n)
+		if err != nil {
+			panic(err)
+		}
+		if preimage.Len() < 32 {
+			return preimage.Bytes()
+		}
+	case emptyNode:
+		return emptyRoot[:]
+	default:
+		panic("Invalid node type")
+	}
+	rlpSize := preimage.Len()
+	d.Write(preimage.Bytes())
+	ret := d.Sum(nil)
+
+	if st.db != nil {
+		if n != nil {
+			st.db.insert(common.BytesToHash(ret), rlpSize, n)
+			// Commit if the size is above 1G
+			if st.db.dirtiesSize > 1073741824 {
```

Comment:
@holiman this line is meant to periodically commit all the dirty cache entries, however I think it's pretty ugly so I'd be happy for some input here.

Reply:
Actually 1G is too much, my machine runs out of memory. 10M works fine and takes 4 times longer.